python/ql/src/semmle/python/frameworks/Stdlib.qll

```diff
@@ -864,6 +864,212 @@ private module Stdlib {
+   * We include the pure paths, as they can be "exported" (say with `as_posix`) and then used to acces the underlying file system.
+   */
+  private string pathlibPathConstructor() {
+    result in ["Path", "PurePath", "PurePosixPath", "PureWindowsPath", "PosixPath", "WindowsPath"]
+  }
+
+  /**
+   * Gets a name of an attribute of a `pathlib.Path` object that is also a `pathlib.Path` object.
+   */
+  private string pathlibPathAttribute() { result in ["parent"] }
+
+  /**
+   * Gets a name of a method of a `pathlib.Path` object that returns a `pathlib.Path` object.
+   */
+  private string pathlibPathMethod() {
+    result in ["absolute", "relative_to", "rename", "replace", "resolve"]
+  }
+
+  /**
+   * Gets a name of a method of a `pathlib.Path` object that modifies a `pathlib.Path` object based on new data.
+   */
+  private string pathlibPathInjection() {
+    result in ["joinpath", "with_name", "with_stem", "with_suffix"]
+  }
+
+  /**
+   * Gets a name of an attribute of a `pathlib.Path` object that exports information about the `pathlib.Path` object.
+   */
+  private string pathlibPathAttributeExport() {
+    result in ["drive", "root", "anchor", "name", "suffix", "stem"]
+  }
+
+  /**
+   * Gets a name of a method of a `pathlib.Path` object that exports information about the `pathlib.Path` object.
+   */
+  private string pathlibPathMethodExport() { result in ["as_posix", "as_uri"] }
+
+  /**
+   * Flow for mehtods that return a `pathlib.Path` object.
+   */
+  private predicate typePreservingCall(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+    exists(DataFlow::AttrRead returnsPath | returnsPath.getAttributeName() = pathlibPathMethod() |
+      nodeTo.(DataFlow::CallCfgNode).getFunction() = returnsPath and
+      nodeFrom = returnsPath.getObject()
+    )
+  }
+
+  /**
+   * Flow for attributes that are `pathlib.Path` objects.
+   */
+  private predicate typePreservingAttribute(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+    exists(DataFlow::AttrRead isPath | isPath.getAttributeName() = pathlibPathAttribute() |
+      nodeTo = isPath and
+      nodeFrom = isPath.getObject()
+    )
+  }
+
+  /**
+   * Gets a reference to a `pathlib.Path` object.
+   * This type tracker makes the monomorphic API use assumption.
+   */
+  private DataFlow::LocalSourceNode pathlibPath(DataFlow::TypeTracker t) {
+    // Type construction
+    t.start() and
+    result = pathlib().getMember(pathlibPathConstructor()).getACall()
+    or
+    // Type-preserving call
+    exists(DataFlow::Node nodeFrom, DataFlow::TypeTracker t2 |
+      nodeFrom.getALocalSource() = pathlibPath(t2) and
+      t2.end()
+    |
+      t.start() and
+      typePreservingCall(nodeFrom, result)
+    )
+    or
+    // Type-preserving attribute access
+    exists(DataFlow::Node nodeFrom, DataFlow::TypeTracker t2 |
+      nodeFrom.getALocalSource() = pathlibPath(t2) and
+      t2.end()
+    |
+      t.start() and
+      typePreservingAttribute(nodeFrom, result)
```

Comment:
Is there a reason (other than a natural desire to keep things separate) to distinguish between `typePreservingCall` and `typePreservingAttribute`?

Perhaps we could simply combine both in `typePreservingStep`?

Reply:
I somehow felt that would become messy, but now I am not so sure. I will give it a try.

Reply:
That turned out fine, I think.